### PR TITLE
fix(keras): align graph mode binary_crossentropy clipping with eager mode (Fixes #118313)

### LINKED ISSUE #118313: [tf.keras][XLA] BinaryCrossentropy(from_logits=False) on explicit sigmoid output returns logits-scale loss under tf.function(jit_compile=True)
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

0.21.0

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04.4 LTS (x86_64)

### Mobile device

Ubuntu 22.04.4 LTS (x86_64)

### Python version

3.10.16

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Description

I found a large numerical discrepancy between eager execution and `tf.function(jit_compile=True)` for `tf.keras.losses.BinaryCrossentropy(from_logits=False)` when the prediction tensor is explicitly computed by `tf.sigmoid(logits)`.

For a large positive logit, `tf.sigmoid(logits)` saturates to `1.0` in float32. In eager mode, passing this saturated probability to `BinaryCrossentropy(from_logits=False)` returns a clipped probability-space loss around `15.94`. However, under `tf.function(jit_compile=True)`, the same expression returns a logits-scale loss around `1449.69`, equal to `BinaryCrossentropy(from_logits=True)(labels, logits)`.

This is surprising because the loss object is explicitly configured with `from_logits=False`, and the value passed to the loss is the result of `tf.sigmoid(logits)`, i.e., a probability tensor.

### Actual behavior

On TensorFlow 2.21.0, float32 policy, GPU execution, I get:
```
Input:
logits = [[1449.6967]]
sigmoid(logits) = [[1.0]]
labels = [[0.0]]

Eager:
BCE(from_logits=False, sigmoid(logits)) = [[15.9423847]]
BCE(from_logits=True, logits)           = [[1449.6967]]

tf.function(jit_compile=True):
BCE(from_logits=False, sigmoid(logits)) = [[1449.6967]]
BCE(from_logits=True, logits)           = [[1449.6967]]
```
The XLA-compiled result for BinaryCrossentropy(from_logits=False) on tf.sigmoid(logits) is exactly equal to the logits-based loss.

The maximal absolute difference between eager and XLA for the probability-input loss is:
```
1433.7542705535889
```

### Standalone code to reproduce the issue

```shell
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import json
import platform
import traceback

import numpy as np
import tensorflow as tf


LOGIT_VALUE = 1449.6966552734375
RTOL = 1e-5
ATOL = 1e-5


def jdump(obj):
    return json.dumps(obj, ensure_ascii=False, sort_keys=True)


def banner(title):
    print("\n" + "=" * 100)
    print(title)
    print("=" * 100)


def to_numpy(x):
    if isinstance(x, tf.Tensor):
        return x.numpy()
    return np.asarray(x)


def summarize(name, x):
    arr = np.asarray(to_numpy(x))
    print(
        f"{name}: shape={arr.shape}, dtype={arr.dtype}, "
        f"value={arr.reshape(-1)[:8].tolist()}"
    )
    return arr


def bce_from_probability_after_sigmoid(logits):
    labels = tf.zeros_like(logits)
    probs = tf.sigmoid(logits)

    loss_fn = tf.keras.losses.BinaryCrossentropy(
        from_logits=False,
        reduction=tf.keras.losses.Reduction.NONE,
    )

    loss = loss_fn(labels, probs)
    return tf.expand_dims(loss, axis=-1)


def bce_from_logits_reference(logits):
    labels = tf.zeros_like(logits)

    loss_fn = tf.keras.losses.BinaryCrossentropy(
        from_logits=True,
        reduction=tf.keras.losses.Reduction.NONE,
    )

    loss = loss_fn(labels, logits)
    return tf.expand_dims(loss, axis=-1)


@tf.function(jit_compile=False)
def graph_bce_from_probability_after_sigmoid(logits):
    return bce_from_probability_after_sigmoid(logits)


@tf.function(jit_compile=True)
def xla_bce_from_probability_after_sigmoid(logits):
    return bce_from_probability_after_sigmoid(logits)


@tf.function(jit_compile=True)
def xla_bce_from_logits_reference(logits):
    return bce_from_logits_reference(logits)


def compare(a, b, name_a, name_b):
    a_np = np.asarray(to_numpy(a))
    b_np = np.asarray(to_numpy(b))

    same_shape = a_np.shape == b_np.shape
    max_abs = float(np.max(np.abs(a_np.astype(np.float64) - b_np.astype(np.float64)))) if same_shape else None
    allclose = bool(np.allclose(a_np, b_np, rtol=RTOL, atol=ATOL, equal_nan=True)) if same_shape else False

    print(f"\nCompare {name_a} vs {name_b}:")
    print(f"  same_shape: {same_shape}")
    print(f"  allclose: {allclose}")
    print(f"  max_abs_diff: {max_abs}")

    return {
        "same_shape": same_shape,
        "allclose": allclose,
        "max_abs_diff": max_abs,
        f"{name_a}_value": a_np.reshape(-1).tolist(),
        f"{name_b}_value": b_np.reshape(-1).tolist(),
    }


def main():
    os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "0")

    tf.keras.mixed_precision.set_global_policy("float32")
    tf.config.run_functions_eagerly(False)
    tf.config.optimizer.set_jit(False)

    try:
        gpus = tf.config.list_physical_devices("GPU")
        if gpus:
            tf.config.set_visible_devices(gpus[0], "GPU")
            tf.config.experimental.set_memory_growth(gpus[0], True)
    except Exception:
        pass

    banner("Environment")
    print("python:", platform.python_version())
    print("platform:", platform.platform())
    print("tensorflow:", tf.__version__)
    print("CUDA_VISIBLE_DEVICES:", os.environ.get("CUDA_VISIBLE_DEVICES", ""))
    print("physical GPUs:", [str(x) for x in tf.config.list_physical_devices("GPU")])
    print("mixed precision policy:", tf.keras.mixed_precision.global_policy())
    print("run_functions_eagerly:", tf.config.functions_run_eagerly())

    logits = tf.constant([[LOGIT_VALUE]], dtype=tf.float32)

    banner("Input")
    summarize("logits", logits)
    summarize("sigmoid(logits)", tf.sigmoid(logits))
    summarize("labels", tf.zeros_like(logits))

    banner("Eager")
    eager_prob_loss = bce_from_probability_after_sigmoid(logits)
    eager_logits_loss = bce_from_logits_reference(logits)

    summarize("eager BCE(from_logits=False, sigmoid(logits))", eager_prob_loss)
    summarize("eager BCE(from_logits=True, logits)", eager_logits_loss)

    banner("tf.function(jit_compile=False) control")
    graph_prob_loss = graph_bce_from_probability_after_sigmoid(logits)
    summarize("graph BCE(from_logits=False, sigmoid(logits))", graph_prob_loss)

    banner("tf.function(jit_compile=True) XLA")
    xla_prob_loss = xla_bce_from_probability_after_sigmoid(logits)
    xla_logits_loss = xla_bce_from_logits_reference(logits)

    summarize("xla BCE(from_logits=False, sigmoid(logits))", xla_prob_loss)
    summarize("xla BCE(from_logits=True, logits)", xla_logits_loss)

    banner("Comparisons")
    cmp_eager_xla = compare(
        eager_prob_loss,
        xla_prob_loss,
        "eager_prob_loss",
        "xla_prob_loss",
    )

    cmp_eager_graph = compare(
        eager_prob_loss,
        graph_prob_loss,
        "eager_prob_loss",
        "graph_prob_loss",
    )

    cmp_xla_prob_logits = compare(
        xla_prob_loss,
        xla_logits_loss,
        "xla_prob_loss",
        "xla_logits_loss",
    )

    banner("FINAL_STATUS")
    payload = {
        "status": "mismatch" if not cmp_eager_xla["allclose"] else "ok",
        "target_check": "eager vs tf.function(jit_compile=True)",
        "op": "tf.keras.losses.BinaryCrossentropy(from_logits=False)(zeros, tf.sigmoid(large_logits))",
        "logit_value": LOGIT_VALUE,
        "eager_vs_xla": cmp_eager_xla,
        "eager_vs_graph_non_xla": cmp_eager_graph,
        "xla_prob_loss_vs_xla_logits_reference": cmp_xla_prob_logits,
    }
    print("FINAL_STATUS=" + jdump(payload))


if __name__ == "__main__":
    try:
        main()
    except Exception:
        print("FAILED")
        traceback.print_exc()
        raise
```

### Relevant log output

```shell
Input:
logits = [[1449.6967]]
sigmoid(logits) = [[1.0]]
labels = [[0.0]]

Eager:
BCE(from_logits=False, sigmoid(logits)) = [[15.9423847]]
BCE(from_logits=True, logits)           = [[1449.6967]]

tf.function(jit_compile=True):
BCE(from_logits=False, sigmoid(logits)) = [[1449.6967]]
BCE(from_logits=True, logits)           = [[1449.6967]]
```

### PR DESCRIPTION
**Pull request by hunterkritik-byte**
### Summary of Changes
Fixes #118313.

This PR resolves a numerical discrepancy between **Eager execution** and **Graph execution / XLA compilation** (`tf.function(jit_compile=True)`) when evaluating `tf.keras.losses.BinaryCrossentropy(from_logits=False)` on inputs produced by `tf.sigmoid(logits)`.

### Root Cause
When `from_logits=False` was passed, `tensorflow/python/keras/backend.py` contained a graph shortcut that checked whether the `output` tensor originated from a `Sigmoid` op (`output.op.type == 'Sigmoid'`). 

If detected, Graph mode unwrapped the sigmoid and routed directly to `nn.sigmoid_cross_entropy_with_logits(labels, logits)`. This bypassed the probability-space epsilon clipping (`clip_by_value(y_pred, epsilon, 1 - epsilon)`), causing extreme saturated logits (e.g., `1449.7`) to evaluate to unclipped logit-space losses (`~1449.7`) in Graph mode while returning clipped probability-space losses (`~15.94`) in Eager mode.

### Solution
- Removed the graph-mode sigmoid unwrapping optimization block in `tensorflow/python/keras/backend.py`. Both Eager and Graph/XLA execution now consistently follow the probability-space epsilon-clipping path when `from_logits=False`.
- Added a regression test `test_binary_crossentropy_eager_graph_consistency` in `tensorflow/python/keras/backend_test.py` to ensure Eager and Graph outputs remain numerically identical under extreme logits.

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for the PR and for tracking down this graph-mode discrepancy!

The logic change to `backend.py` looks correct, but I noticed two issues preventing CI from running and passing on your new test file:

1. **Missing `BUILD` Target**: The new test file `tensorflow/python/keras/backend_test.py` needs to be added to `tensorflow/python/keras/BUILD` using a `tf_py_test` (or `py_test`) rule. Currently, Bazel ignores the new file which is why the core tests falsely reported success.
2. **Missing Copyright Header**: `backend_test.py` needs the standard TensorFlow Apache 2.0 license copyright header at the top of the file, which causes the `License Check` CI to fail.

Could you please add the file to the `BUILD` configuration and insert the copyright header? We can re-review once those are passing!
